base/array.jl

```diff
@@ -656,6 +656,20 @@ function collect_to_with_first!(dest, v1, itr, st)
     return grow_to!(dest, itr, st)
 end
 
+function setindex_widen!(dest::Array{T}, el, i) where T
+    @_inline_meta
+    if el isa T || typeof(el) === T
+        @inbounds dest[i] = el::T
+        dest
+    else
+        R = promote_typejoin(T, typeof(el))
+        new = similar(dest, R)
+        copyto!(new,1, dest,1, i-1)
```

Comment:
If this were more generic, for use for out-of-order setindexing by end users or offset arrays or whatever, we’d probably need to copy the whole array. Pity.

Reply:
Yeah the 1's tipped me off this wouldn't work for OffsetArrays

Reply:
Should I add an AbstractArray method which just copies the whole thing?

Reply:
I'd rather use `firstindex`.

Reply:
Ok done